### PR TITLE
redux, redux-saga were added

### PR DESCRIPTION
**Pull request by dmytrocherepakhin**

### CODE REVIEW

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 1]
```diff
@@ -28,32 +46,10 @@ export interface IComics {
...
-    totalOfItems: number
-}
-
-class ComicsPage extends React.Component<IProps, IState> {
-    constructor(props: IProps) {
+class ComicsPage extends React.Component<IComicsPops> {
+    constructor(props: IComicsPops) {
         super(props)
         this.setCurrentPage = this.setCurrentPage.bind(this)
-        this.state = {
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx:48]: you don't need to move whole state to redux, only data that you gets from api and loading state

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 2]
```diff
@@ -6,20 +6,38 @@ import ProgressBarIndeterminate from '../ProgressBarIndeterminate';
...
+    progressBar: boolean,
+    heroName: string,
+    comics: IComics[],
+    totalOfItems: number,
+    setComicsCurrentPage(currentComicsPage: number): ISetComicsCurrentPage,
+    fetchComics(comics: IComics[], totalOfItems: number): IFetchComics,
+    fetchComicsHero(heroName: string): IFetchComicsHero,
+    hideLoader(): ILoader,
+    showLoader(): ILoader,
+    fetchComicsSaga(page: string | string[] | null, heroId: number, currentComicsPage: number): IFetchComicsSaga
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: too many types for page param

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 3]
```diff
@@ -6,20 +6,38 @@ import ProgressBarIndeterminate from '../ProgressBarIndeterminate';
...
     history: RouteComponentProps["history"],
+    currentComicsPage: number,
+    progressBar: boolean,
+    heroName: string,
+    comics: IComics[],
+    totalOfItems: number,
+    setComicsCurrentPage(currentComicsPage: number): ISetComicsCurrentPage,
+    fetchComics(comics: IComics[], totalOfItems: number): IFetchComics,
+    fetchComicsHero(heroName: string): IFetchComicsHero,
+    hideLoader(): ILoader,
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: loader methods should be in react

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 4]
```diff
@@ -6,20 +6,38 @@ import ProgressBarIndeterminate from '../ProgressBarIndeterminate';
...
 import queryString from 'query-string';
 import { RouteComponentProps } from "react-router-dom";
-import { getComics } from '../api'
-import { getComicsHero } from '../api'
+import { connect } from 'react-redux'
+import { fetchComics, fetchComicsHero, fetchComicsSaga, hideLoader, IFetchComics, IFetchComicsHero, IFetchComicsSaga, ILoader, ISetComicsCurrentPage, setComicsCurrentPage, showLoader } from "../../store/actions/actions";
+import { AppDispatch, RootState } from "../HomePage/HomePage";
 
-type IProps = {
+export type IComicsPops = {
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: interface consists of field related to different entities, please split it by sense

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 5]
```diff
@@ -6,20 +6,38 @@ import ProgressBarIndeterminate from '../ProgressBarIndeterminate';
 import PaginationRounded from '../PaginationRounded';
 import queryString from 'query-string';
 import { RouteComponentProps } from "react-router-dom";
-import { getComics } from '../api'
-import { getComicsHero } from '../api'
+import { connect } from 'react-redux'
+import { fetchComics, fetchComicsHero, fetchComicsSaga, hideLoader, IFetchComics, IFetchComicsHero, IFetchComicsSaga, ILoader, ISetComicsCurrentPage, setComicsCurrentPage, showLoader } from "../../store/actions/actions";
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: long string, eslint should show error or at least warning in this case

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 6]
```diff
@@ -100,18 +89,40 @@ class ComicsPage extends React.Component<IProps, IState> {
...
+        progressBar: state.loaderReducer.progressBar,
+        heroName: state.comicsReducer.heroName,
+        comics: state.comicsReducer.comics,
+        totalOfItems: state.comicsReducer.totalOfItems
+    };
+}
+
+const mapDispatchToProps = (dispatch: AppDispatch) => {
+    return {
+        setComicsCurrentPage: (currentComicsPage: number) => dispatch(setComicsCurrentPage(currentComicsPage)),
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: should be handled by react

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 7]
```diff
@@ -62,32 +58,25 @@ class ComicsPage extends React.Component<IProps, IState> {
...
-        const comicsResult = await getComics(this.state, this.props);
-        this.setState({
-            heroName: ComicsHeroResult.data.data.results[0].name,
-            comics: comicsResult.data.data.results,
-            totalOfItems: comicsResult.data.data.total,
-            progresBar: false
-        });
+        const page = queryString.parse(this.props.location.search).page;
+        const heroId = this.props.match.params.id;
+        const currentComicsPage = this.props.currentComicsPage;
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: you can use destructuring to get props value and not to write this.props... all the time

Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: currentPage also better to leave in state

File: marvel_superheroes/src/components/HomePage/HomePage.tsx [hunk 8]
```diff
@@ -7,7 +7,9 @@ import queryString from 'query-string';
 import { RouteComponentProps } from "react-router";
 import HeroSearchBar from "../HeroSearchBar/HeroSearchBar";
 import PaginationRounded from '../PaginationRounded';
-import { getHeroes } from "../api";
+import { connect } from 'react-redux'
+import { fetchHeroes, fetchHeroesSaga, hideLoader, IFetchHeroes, IFetchHeroesSaga, ILoader, ISetHeroesCurrentPage, ISetHeroesNameStartWith, ISetHeroesOrderBy, setHeroesCurrentPage, setHeroesNameStartWith, setHeroesOrderBy, showLoader } from "../../store/actions/actions";
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/HomePage/HomePage.tsx]: the same about linter

File: marvel_superheroes/src/components/HomePage/HomePage.tsx [hunk 9]
```diff
@@ -7,7 +7,9 @@ import queryString from 'query-string';
...
+import { store } from "../../index";
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/HomePage/HomePage.tsx]: also see that some strings are double quote  and others are single. Add appropriate eslint rule to handle this

File: marvel_superheroes/src/components/HomePage/HomePage.tsx [hunk 10]
```diff
@@ -17,41 +19,40 @@ export interface IHero {
...
-    limit: number,
     progressBar: boolean,
+    currentHeroesPage: number,
+    setHeroesNameStartWith(nameStartsWith: string): ISetHeroesNameStartWith,
+    setHeroesOrderBy(orderBy: string): ISetHeroesOrderBy,
+    setHeroesCurrentPage(currentHeroesPage: number): ISetHeroesCurrentPage,
+    fetchHeroes(heroes: IHero[], totalOfItems: number): IFetchHeroes,
+    hideLoader(): ILoader,
+    showLoader(): ILoader,
+    fetchHeroesSaga(orderBy: string, currentHeroesPage: number, query: string | string[] | null, sort: string | string[] | null, page: string | string[] | null): IFetchHeroesSaga
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/HomePage/HomePage.tsx]: parameters shouldn't have so many union types

File: marvel_superheroes/src/components/HomePage/HomePage.tsx [hunk 11]
```diff
@@ -17,41 +19,40 @@ export interface IHero {
...
+interface IGetHeroes {
+    heroes: IHero[],
+    totalOfItems: number,
     nameStartsWith: string,
     orderBy: string,
-    limit: number,
     progressBar: boolean,
+    currentHeroesPage: number,
+    setHeroesNameStartWith(nameStartsWith: string): ISetHeroesNameStartWith,
+    setHeroesOrderBy(orderBy: string): ISetHeroesOrderBy,
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/HomePage/HomePage.tsx]: all the same about this interface, no need to pass all methods and state to redux. There is no sense in moving setHeroesNameStartWith, setHeroesOrderBy, setHeroesCurrentPage and so on

File: marvel_superheroes/src/components/HomePage/HomePage.tsx [hunk 12]
```diff
@@ -66,52 +67,43 @@ class HomePage extends React.Component<IProps, IState> {
...
     async componentDidMount(): Promise<void> {
         this.makeRequest()
     }
 
-    async componentDidUpdate(prevProps: IProps, prevState: IState): Promise<void> {
-        if (this.props.location !== prevProps.location || this.state.offset !== prevState.offset) {
-            this.makeRequest()
-        }
+    async componentDidUpdate(prevProps: IHeroProps): Promise<void> {
+        if (this.props.location !== prevProps.location || this.props.currentHeroesPage !== prevProps.currentHeroesPage) { this.makeRequest(); }
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/HomePage/HomePage.tsx]: code stye...

File: marvel_superheroes/src/components/api.tsx [hunk 13]
```diff
@@ -1,17 +1,15 @@
...
-    const limit = requestData.limit;
-    const orderBy = requestData.orderBy;
-    const offset = requestData.offset;
-    const { query, sort, page } = queryString.parse(props.location.search);
+export async function getHeroes(
+    orderBy: string,
+    currentHeroesPage: number,
+    query: string | string[] | null,
+    sort: string | string[] | null,
+    page: string | string[] | null): Promise<AxiosResponse> {
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/api.tsx]: same problem with types

File: marvel_superheroes/src/store/actionTypes/types.tsx [hunk 14]
```diff
@@ -0,0 +1,15 @@
+export const SET_HEROES_NAME_START_WITH = 'SET_HEROES_NAME_START_WITH'
```
Comment (alinalaktionova) [inline marvel_superheroes/src/store/actionTypes/types.tsx]: You should have only 3 actions for each backend request. For example, GET_HERO_REQUEST, GET_HERO_SUCCESS, GET_HERO_ERROR. Rewrite all these actions

File: marvel_superheroes/src/store/reducers/comicsReducer.tsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+                ...state,
+                comics: action.comics,
+                totalOfItems: action.totalOfItems
+            }
+        case FETCH_COMICS_HERO:
+            return {
+                ...state,
+                heroName: action.heroName
+            }
+        case FETCH_COMICS_SAGA:
```
Comment (alinalaktionova) [inline marvel_superheroes/src/store/reducers/comicsReducer.tsx]: unusefull case

File: marvel_superheroes/src/store/reducers/heroesReducer.tsx [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+                ...state,
+                currentHeroesPage: action.currentHeroesPage
+            };
+        case FETCH_HEROES:
+            return {
+                ...state,
+                heroes: action.heroes,
+                totalOfItems: action.totalOfItems
+            };
+        case FETCH_HEROES_SAGA:
```
Comment (alinalaktionova) [inline marvel_superheroes/src/store/reducers/heroesReducer.tsx]: same unuseful

File: marvel_superheroes/src/store/sagas/comicsSaga.tsx [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    try {
+        yield put(showLoader())
+        const comicsHeroResult: AxiosResponse = yield call(getComicsHero, payload.heroId)
+        const comicsResult: AxiosResponse = yield call(getComics, payload.currentComicsPage, payload.page, payload.heroId)
+        yield put(fetchComicsHero(comicsHeroResult.data.data.results[0].name))
+        yield put(fetchComics(comicsResult.data.data.results, comicsResult.data.data.total))
+        yield put(hideLoader())
+    } catch (e) {
+        yield (console.log('Error'))
+        yield put(hideLoader())
```
Comment (alinalaktionova) [inline marvel_superheroes/src/store/sagas/comicsSaga.tsx]: should add errorAction in catch block

**Review by alinalaktionova — CHANGES_REQUESTED** [review]

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx (hunk 2 above)
Comment (dmytrocherepakhin) [reply marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx:31]: the page param was received as 'queryString.parse(this.props.location.search).page'  with only type 'string | string[] | null'  ??

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 18]
```diff
@@ -1,128 +1,130 @@
...
+  addressBarMaker = (pageArg: number | null): void => {
+    this.props.history.push("?page=" + pageArg)
+  }
 
-    setCurrentPage = (currentComicsPage: number): void => {
-        this.props.setComicsCurrentPage(currentComicsPage);
-        this.addressBarMaker(currentComicsPage);
-    }
+  setCurrentPage = (currentComicsPage: number): void => {
+    this.setState({ currentComicsPage: currentComicsPage })
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: you can write just this.setState({ currentComicsPage }), it's the same

File: marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx [hunk 19]
```diff
@@ -1,128 +1,130 @@
...
 const mapStateToProps = (state: RootState) => {
-    return {
-        currentComicsPage: state.comicsReducer.currentComicsPage,
-        progressBar: state.loaderReducer.progressBar,
-        heroName: state.comicsReducer.heroName,
-        comics: state.comicsReducer.comics,
-        totalOfItems: state.comicsReducer.totalOfItems
-    };
+  return {
+    heroName: state.comicsReducer.heroName,
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/ComicsPage/ComicsPage.tsx]: can use destructuring for this

File: marvel_superheroes/src/components/Hero/Hero.tsx [hunk 20]
```diff
@@ -1,31 +1,38 @@
-import React from "react";
-import { Link } from "react-router-dom";
-import './Hero.css';
-import { IHero } from "../HomePage/HomePage";
+import React from "react"
+import { Link } from "react-router-dom"
+import "./Hero.css"
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/Hero/Hero.tsx]: better to use single quote and semicolon in the end of statements. change eslint rules

File: marvel_superheroes/src/components/HomePage/HomePage.tsx [hunk 21]
```diff
@@ -1,156 +1,154 @@
...
-        if (nameStartsWith !== queryIsExist) { this.props.setHeroesNameStartWith(query) }
-        if (orderBy !== sortIsExist) { this.props.setHeroesOrderBy(sort) }
-        this.props.setHeroesCurrentPage(1)
+  }
+
+  render(): JSX.Element {
+    const heroes = { heroes: this.props.heroes }
+    const queryStringParse = queryString.parse(this.props.location.search)
+    const query = queryStringParse.query?.toString()
+    const sort = queryStringParse.sort?.toString()
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/HomePage/HomePage.tsx]: use descructuring

File: marvel_superheroes/src/components/api.tsx [hunk 22]
```diff
@@ -1,97 +1,81 @@
...
-    try {
-        return requestHeroes;
-    } catch (err) {
-        throw new Error('Error  of requestHeroes');
-    }
-}
+  const params: Params = {
+    limit: limit,
+    offset: offset,
+    orderBy: orderBy,
```
Comment (alinalaktionova) [inline marvel_superheroes/src/components/api.tsx]: again you don't need to duplicate property if name is the same

**Review by alinalaktionova — CHANGES_REQUESTED** [review]

**Review by alinalaktionova — APPROVED** [review]

**Pull request merged**